[DX-1310] Add docs for the custom rate limit keys (#4563)
TykTechnologies/tyk-docs: tyk-docs/content/tyk-stack/tyk-developer-portal/enterprise-developer-portal/api-access/configuring-custom-rate-limit-keys.md

---
title: "Configuring custom rate limit keys"
date: 2022-02-11
tags: ["Tyk Developer Portal","Enterprise Portal", "Rate limit", "Quota"]
description: "How to configure custom rate limit keys with the Enterprise Portal"
menu:
  main:
    parent: "API Access"
weight: 5
---

{{< note success >}}
**Tyk Enterprise Developer Portal**

If you are interested in getting access contact us at [support@tyk.io](<mailto:support@tyk.io?subject=Tyk Enterprise Portal Beta>)

{{< /note >}}

## Introduction

> Different business models may require applying rate limits and quotas not only by credentials but also by other entities, such as per application, per developer, per organization, and so on.

Reviewer: ```suggestion
Different business models may require applying rate limits and quotas not only by credentials but also by other entities, e.g. per application, per developer, per organization etc.
```

> For instance, if an API Product is sold to a B2B customer, the quota of API calls is usually applied to all developers and their respective applications combined, and to a specific credential.

Reviewer: ```suggestion
For example, if an API Product is sold to a B2B customer, the quota of API calls is usually applied to all developers and their respective applications combined, in addition to a specific credential.
```

> 

Reviewer: remove additional blank line

Author: Done

To enable this, Tyk introduced support for custom rate limit keys in [Tyk 5.3.0]({{< ref "product-stack/tyk-dashboard/release-notes/version-5.3" >}}). This guide explains how to configure custom rate limit keys.

## Prerequisites for getting started
This capabilty works with [Tyk 5.3.0]({{< ref "product-stack/tyk-dashboard/release-notes/version-5.3" >}}) or higher. 

## Configuring custom rate limit keys for policies in Tyk Dashboard

> Custom rate limit keys are applied at a policy level. When a custom rate limit key is specified, quota, rate limit, and throttling will be calculated against the specified value and not against a credential ID.

Reviewer: ```suggestion
Custom rate limit keys are applied at a policy level. When a custom rate limit key is specified, quota, rate limit and throttling will be calculated against the specified value and not against a credential ID.
```

To specify a custom rate limit key, add to a policy a new metadata field called `rate_limit_pattern`. In the value field you can specify any value or expression that you want to use as a custom rate limit key for your APIs.
The `rate_limit_pattern` syntax support referencing session metadata using `$tyk_meta.FIELD_NAME` reference. In addition, it's possible to concatenate multiple values together using the pipe operator (`|`).

> For instance, to specify a rate limit pattern to calculate rate limit for a combination of developers and plans (thus, all credentials of a developer that use the same plan would share the same rate limit) you can use the following expression (assuming the `DeveloperID` and `PlanID` metadata fields are available in a session):

Reviewer: maybe reduce the length of the sentence. The bracketed sentence makes the sentence too long and difficult to read

Author: Done

```gotemplate
$tyk_meta.DeveloperID|$tyk_meta.PlanID
```

{{< img src="/img/dashboard/portal-management/enterprise-portal/configuring-custom-rate-limit-keys.png" alt="Configuring custom rate limit keys" >}}

{{< note success >}}
**Updating credential metadata**

Please note that the custom rate limit key capability uses only metadata objects, such as credentials metadata available in a session.
Therefore, if the `rate_limit_pattern` relies on credentials metadata, this capability will work only if those values are present.
If, after evaluating the `rate_limit_pattern`, its value is equal to an empty string, the rate limiter behavior defaults to rate limiting by credential IDs.

{{< /note >}}

## Using custom rate limit keys with the portal

The Tyk Enterprise Developer Portal facilitates the configuration of various rate limiting options based on a business model for API Products published in the portal.

> To achieve this, the portal, by default, populates the following attributes in the credential metadata:

Reviewer: Add a blank line before this line, i.e. before line 55

Author: Done

- **ApplicationID**: The ID of the application to which the credential belongs.
- **DeveloperID**: The ID of the developer who created the credential.
- **OrganisationID**: The ID of the organisation to which the developer belongs.
- **TeamIDs**: A list of team IDs to which the developer belongs.

Additionally, it's possible to attach [custom attribute values]({{< ref "product-stack/tyk-enterprise-developer-portal/portal-customisation/customise-user-model#add-attributes-to-the-user-model" >}}) defined in a developer profile as metadata fields to credentials.

When a credential is provisioned by the portal, all the fields described above are added as metadata values to the credential, making them valid options for configuring the rate limit key:

> {{< img src="/img/dashboard/portal-management/enterprise-portal/credential-metadata.png" alt="Credential's metadata" >}}

Reviewer: Add blank line before the img

Author: Done

> 

Reviewer: Remove additional blank line

> This approach allows the portal to seamlessly apply rate limits based on any combination of the aforementioned fields and other custom metadata objects defined in policies used for plans or products, or on credentials.

Reviewer: ```suggestion
This approach allows the portal to seamlessly apply rate limits based on any combination of the aforementioned fields and other custom metadata objects defined in policies used for plans or products. This is in addition to credentials.
```

To specify a custom rate limit key, add to a policy a new metadata field called `rate_limit_pattern`.
In the value field you can specify any value or expression that you want to use as a custom rate limit key for your APIs.
The `rate_limit_pattern` syntax support referencing session metadata using `$tyk_meta.FIELD_NAME` reference.
In addition, it's possible to concatenate multiple values together using the pipe operator (`|`).

For instance, if you want to specify a rate limit pattern to calculate the rate limit for a combination of developers and plans, where all credentials of a developer using the same plan share the same rate limit, you can use the following expression.
This assumes that the `DeveloperID` and `PlanID` metadata fields are available in a session:

```gotemplate
$tyk_meta.DeveloperID|$tyk_meta.PlanID
```

Here's how it looks like in the Tyk Dashboard UI:

{{< img src="/img/dashboard/portal-management/enterprise-portal/configuring-custom-rate-limit-keys.png" alt="Configuring custom rate limit keys" >}}

> 

Author: Fixed

{{< note success >}}
**Updating credential metadata**

Please note that the custom rate limit key capability uses only metadata objects, such as credentials metadata available in a session.
Therefore, if the `rate_limit_pattern` relies on credentials metadata, this capability will work only if those values are present.
If, after evaluating the `rate_limit_pattern`, its value is equal to an empty string, the rate limiter behavior defaults to rate limiting by credential IDs.

{{< /note >}}

## Using custom rate limit keys with the portal

The Tyk Enterprise Developer Portal facilitates the configuration of various rate limiting options based on a business model for API Products published in the portal.

To achieve this, the portal, by default, populates the following attributes in the credential metadata:
- **ApplicationID**: The ID of the application to which the credential belongs.
- **DeveloperID**: The ID of the developer who created the credential.
- **OrganisationID**: The ID of the organisation to which the developer belongs.

> - **TeamIDs**: A list of team IDs to which the developer belongs.

Author: We don't track ratelimits for individual team - the RL key is always a string. So if a developer belongs to multiple teams, the RL key will a combination of IDs of those team. We populate this field in the credentials metadata for other purposes e.g. sending it to 3rd party systems through custom plugins. I will remove that for now so that developers won't be confused.
